src/code/providers/provider-interface.ts

```diff
@@ -0,0 +1,550 @@
+    for (let key in envelopeMetadata) {
+      (this.envelopeMetadata as any)[key] = (this.envelopeMetadata as any)[key]
+    }
+  }
+
+  // returns new CloudContent containing enveloped data
+  createEnvelopedCloudContent(content: CloudContent) {
+    return new CloudContent((this.envelopContent(content)), (this._identifyContentFormat(content)))
+  }
+
+  // envelops content with metadata, returns an object.
+  // If content was already an object (Object or JSON) with metadata,
+  // any existing metadata will be retained.
+  // Note: calling `envelopContent` may be safely called on something that
+  // has already had `envelopContent` called on it, and will be a no-op.
+  envelopContent(content: CloudContent) {
+    const envelopedCloudContent:any = this._wrapIfNeeded(content)
+    for (let key in this.envelopeMetadata) {
+      if (envelopedCloudContent[key] == null) {
+        envelopedCloudContent[key] = (this.envelopeMetadata as any)[key]
+      }
+    }
+    return envelopedCloudContent
+  }
+
+  _identifyContentFormat(content: CloudContent | string) {
+    if ((content == null)) { return }
+    const result = { isCfmWrapped: false, isPreCfmFormat: false }
+    if (isString(content)) {
+      try { content = JSON.parse(content as string) } catch (error) {
+        // noop, just cecking if it's valid json
+      }
+    }
+    // Currently, we assume 'metadata' is top-level property in
+    // non-CFM-wrapped documents. Could put in a client callback
+    // that would identify whether the document required
+    // conversion to eliminate this assumption from the CFM.
+    if ((content as CloudContent).metadata) {
+      return result
+    }
+    if (
+        // 'cfmWrapped' means meta-data is top-level, and content
+        // can be found inside content.content
+        ((content as CloudContent).cfmVersion != null) ||
+        ((content as CloudContent).content != null)) {
+      result.isCfmWrapped = true
+    } else {
+      result.isPreCfmFormat = true
+    }
+    return result
+  }
+
+  // envelops content in {content: content} if needed, returns an object
+  _wrapIfNeeded(content: string | CloudContent) {
+    if (isString(content)) {
+      try { content = JSON.parse(content as string) } catch (error) {
+        // noop, just cecking if it's json or plain text
+      }
+    }
+    if ((content as CloudContent).content != null) {
+      return content
+    } else {
+      return {content}
+    }
+  }
+}
+
+interface ICloudContentFormat {
+  isCfmWrapped: boolean
+  isPreCfmFormat: boolean
+}
+
+interface ICloudContentInterface {
+  contentFormat: ICloudContentFormat
+  content: any
+  // If we are 'wrapped' data & meta-data are both top-level
+  [key: string]: any
+}
+class CloudContent implements ICloudContentInterface{
+  static wrapFileContent: boolean = true
+
+  content : any
+  contentFormat: any
+  [key: string]: any
+  constructor(_content:any, _contentFormat:any) {
```

Comment:
I think you can rename `_content` to `content` so it doesn't look like a private variable, same with `_contentFormat`.